automata/fa/fa.py

```diff
@@ -12,3 +20,217 @@ class FA(Automaton, metaclass=abc.ABCMeta):
     """An abstract base class for finite automata."""
 
     __slots__ = tuple()
+
+    @staticmethod
+    def get_state_name(state_data: FAStateT) -> str:
+        """
+        Get an string representation of a state. This is used for displaying and
+        uses `str` for any unsupported python data types.
+        """
+        if isinstance(state_data, str):
+            if state_data == "":
+                return "λ"
+
+            return state_data
+
+        if isinstance(state_data, (set, frozenset, list, tuple)):
+            inner = ", ".join(FA.get_state_name(sub_data) for sub_data in state_data)
+            if isinstance(state_data, (set, frozenset)):
+                if state_data:
+                    return "{" + inner + "}"
+                else:
+                    return "∅"
+
+            if isinstance(state_data, tuple):
+                return "(" + inner + ")"
+
+            if isinstance(state_data, list):
+                return "[" + inner + "]"
+
+        return str(state_data)
+
+    def get_edge_name(self, symbol) -> str:
+        return str(symbol)
+
+    @abc.abstractmethod
+    def iter_states(self) -> Iterable[FAStateT]:
+        """Iterate over all states in the automaton."""
+
+    @abc.abstractmethod
+    def iter_transitions(self) -> Iterable[tuple[FAStateT, FAStateT, Any]]:
+        """
+        Iterate over all transitions in the automaton. Each transition is a tuple
+        of the form (from_state, to_state, symbol)
+        """
+
+    @abc.abstractmethod
+    def is_accepting(self, state: FAStateT) -> bool:
+        """Check if a state is an accepting state."""
+
+    @abc.abstractmethod
+    def is_initial(self, state: FAStateT) -> bool:
```

Comment:
@khoda81 I agree with @eliotwrobson. I would prefer to drop this method for this PR.

Reply:
@eliotwrobson is it okay to turn them into abstract properties instead?

Reply:
@eliotwrobson Is it okay to turn them into abstract properties instead?

Reply:
The use of `abstractproperty` is actually deprecated (see https://docs.python.org/3/library/abc.html#abc.abstractproperty). I think it's unnecessary since these are always set in the `__init__` function at no cost, so there's not really a benefit to making them properties.